File: 3427-special-array-ii/3427-special-array-ii.js
/**
 * @param {number[]} nums
 * @param {number[][]} queries
 * @return {boolean[]}
 */
var isArraySpecial = function(nums, queries) {
    var n = nums.length, q = queries.length;
    const help = Array(n).fill(0)

    for (let i = 1; i < n; ++i)
        help[i] = (((nums[i] ^ nums[i - 1]) & 1) != 0) ? help[i - 1] : i;
    
    const ans = Array(q).fill(false);
    for (let i = 0; i < q; ++i) {
        let f = queries[i][0];
        let t = queries[i][1];
        ans[i] = help[t] <= f;
    }
    return ans;
};

File: 3427-special-array-ii/3427-special-array-ii.go
func isArraySpecial(nums []int, queries [][]int) []bool {
    n, q := len(nums), len(queries)

    help := make([]int, n)
    for i := 1; i < n; i++ {
        if (nums[i] ^ nums[i - 1]) & 1 != 0 {
            help[i] = help[i - 1]
        } else {
            help[i] = i
        }
    }

    ans := make([]bool, q)
    for i := 0; i < q; i++ {
        f := queries[i][0]
        t := queries[i][1]
        ans[i] = help[t] <= f
    }
    return ans
}